src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelLoopState.cs

```diff
@@ -0,0 +1,628 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+
+//
+// A non-generic and generic parallel state class, used by the Parallel helper class
+// for parallel loop management.
+//
+// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+
+// Prevents compiler warnings/errors regarding the use of ref params in Interlocked methods
+
+#pragma warning disable 0420
+namespace System.Threading.Tasks
+{
+    /// <summary>
+    /// Enables iterations of <see cref="T:System.Threading.Tasks.Parallel"/> loops to interact with
+    /// other iterations.
+    /// </summary>
+    [DebuggerDisplay("ShouldExitCurrentIteration = {ShouldExitCurrentIteration}")]
+    public class ParallelLoopState
+    {
+        // Derived classes will track a ParallelStateFlags32 or ParallelStateFlags64.
+        // So this is slightly redundant, but it enables us to implement some 
+        // methods in this base class.
+        private ParallelLoopStateFlags _flagsBase;
+
+        internal ParallelLoopState(ParallelLoopStateFlags fbase)
+        {
+            _flagsBase = fbase;
+        }
+
+        /// <summary>
+        /// Internal/virtual support for ShouldExitCurrentIteration.
+        /// </summary>
+        internal virtual bool InternalShouldExitCurrentIteration
+        {
+            get
+            {
+                Contract.Assert(false);
```

Comment:
@ellismg recently swept through all of the code in GitHub to change Contract.Asserts over to Debug.Assert and Debug.Fail calls.